File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/classical/urgency/Sweep.java
package cujae.inf.ic.om.heuristic.assignment.classical.urgency;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.factory.DistanceType;
import cujae.inf.ic.om.matrix.NumericMatrix;

public class Sweep extends ByUrgency implements IUrgencyWithMU {
	
	public static DistanceType distanceType = DistanceType.Real;
	private Solution solution = new Solution();
	
	private ArrayList<Cluster> listClusters;	
	private ArrayList<Customer> listCustomersToAssign;
	private ArrayList<Integer> listIDDepots;
	
	private NumericMatrix urgencyMatrix;
	private NumericMatrix closestMatrix;
	
	private ArrayList<ArrayList<Integer>> listDepotsOrdered;
	private int muIDDepot;
	private ArrayList<Double> listUrgencies;

	public Sweep() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
	
	@Override
	public void initialize() {
		listClusters = initializeClusters();	
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		
		listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());
		
		urgencyMatrix = initializeCostMatrix(listCustomersToAssign, Problem.getProblem().getDepots(), distanceType);
		closestMatrix = new NumericMatrix(urgencyMatrix);
		
		listDepotsOrdered = getDepotsOrdered(listCustomersToAssign, listIDDepots, closestMatrix);
		muIDDepot = findClusterWithMU(listClusters);
		listUrgencies = getListUrgencies(listCustomersToAssign, listDepotsOrdered, urgencyMatrix, muIDDepot);	
	}
	
	@Override
	public void assign() {
		int posCustomer = -1;
		int idCustomer = -1;
		double requestCustomer = 0.0;
		
		int idClosestDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
		
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty())) 
		{
			posCustomer = getPosMaxValue(listUrgencies);
			idCustomer = listCustomersToAssign.get(posCustomer).getIDCustomer();
			requestCustomer = listCustomersToAssign.get(posCustomer).getRequestCustomer();
			
			idClosestDepot = listDepotsOrdered.get(posCustomer).get(0).intValue(); 
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idClosestDepot));
			
			posCluster = findCluster(idClosestDepot, listClusters);

			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();

				if(capacityDepot >= (requestCluster + requestCustomer))
				{
					requestCluster += requestCustomer;
					
					listClusters.get(posCluster).setRequestCluster(requestCluster);
					listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

					listCustomersToAssign.remove(posCustomer);
					listUrgencies.remove(posCustomer);
					listDepotsOrdered.remove(posCustomer);
					
					if(idClosestDepot == muIDDepot)
					{
						muIDDepot = findClusterWithMU(listClusters);

						if(idClosestDepot != muIDDepot)
						{
							urgencyMatrix = initializeCostMatrix(Problem.getProblem().getCustomers(), Problem.getProblem().getDepots(), distanceType);
							listUrgencies = getListUrgencies(listCustomersToAssign, listDepotsOrdered, urgencyMatrix, muIDDepot);
						}
					}
				}
				else
				{
					if(capacityDepot > requestCluster)
					{
						listDepotsOrdered.get(posCustomer).remove(0);
	
						if(listDepotsOrdered.get(posCustomer).isEmpty())
						{
							solution.getUnassignedItems().add(idCustomer);

							listCustomersToAssign.remove(posCustomer);
							listUrgencies.remove(posCustomer);
							listDepotsOrdered.remove(posCustomer);							
						}
						else
							listUrgencies.set(posCustomer, getUrgency(idCustomer, listDepotsOrdered.get(posCustomer), urgencyMatrix, muIDDepot));
					}						
					else
					{					
						int posDepot = Problem.getProblem().findPosElement(listIDDepots, idClosestDepot);
						int currentPosDepot = -1;
						int totalAvailableDepots = 0;
						
						for(int i = 0; i < listDepotsOrdered.size(); i++)
						{ 
							currentPosDepot = Problem.getProblem().findPosElement(listDepotsOrdered.get(i), idClosestDepot);

							if(currentPosDepot != -1)
							{
								totalAvailableDepots = listDepotsOrdered.get(i).size();

								if(totalAvailableDepots == 1)
								{
									solution.getUnassignedItems().add(listCustomersToAssign.get(i).getIDCustomer());
									
									listCustomersToAssign.remove(i);
									listUrgencies.remove(i);
									listDepotsOrdered.remove(i);								
								}
								else
								{
									if(currentPosDepot == 0)
										listUrgencies.set(i, getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listDepotsOrdered.get(i), urgencyMatrix, muIDDepot));
									
									listDepotsOrdered.get(i).remove(currentPosDepot);
								}	
							}
						}		
						listIDDepots.remove(posDepot);
						
						if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(listClusters.remove(posCluster));
						else
							listClusters.remove(posCluster);
					}
				}
			}
		}
	}
	
	@Override
	public Solution finish() {
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));

		OSRMService.clearDistanceCache();
		
		return solution;
	}
	
	/**
     * @param ArrayList<Cluster> listado de clusteres
     * @return int posición del cluster de mayor demanda insatisfecha
     * Retorna el identificador del depósito cuyo cluster es el de mayor demanada insatisfecha 
     **/
	private int findClusterWithMU(ArrayList<Cluster> clusters){
		int idDepot = clusters.get(0).getIDCluster();
		double muRequest = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idDepot)) - clusters.get(0).getRequestCluster();
		
		double cuRequest = 0.0;
		
		for(int i = 1; i < clusters.size(); i++)
		{
			cuRequest = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(clusters.get(i).getIDCluster())) - clusters.get(i).getRequestCluster();
			
			if(cuRequest > muRequest)
			{
				muRequest = cuRequest;
				idDepot = clusters.get(i).getIDCluster();
			}						
		}
		return idDepot;
	}
	
	 /**
     * @param  ArrayList<Customer> listado de clientes
     * @param  ArrayList<Integer> listado de identificadores de depósitos
     * @param  NumericMatrix matriz con las distancias
     * @param  int identificador del depósito de mayor demanda insatisfecha
     * @return ArrayList<Double> listado de urgencia
     * Retorna un listado con las urgencias de los clientes del listado entrado por parámetro
     **/
	public ArrayList<Double> getListUrgencies(ArrayList<Customer> listCustomersToAssign, ArrayList<ArrayList<Integer>> listIDDepots, NumericMatrix urgencyMatrix, int muIDDepot){
		ArrayList<Double> urgencies = new ArrayList<Double>();
		
		if(listIDDepots.size() > 1)
			for(int i = 0; i < listCustomersToAssign.size(); i++)
				urgencies.add(getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listIDDepots.get(i), urgencyMatrix, muIDDepot));
		else
			for(int i = 0; i < listCustomersToAssign.size(); i++)
				urgencies.add(getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listIDDepots.get(0), urgencyMatrix, muIDDepot));

		return urgencies;
	}

	/**
     * @param  int identificador del cliente
     * @param  ArrayList<Integer> listado de identificadores de depósitos
     * @param  NumericMatrix matriz con las distancias
     * @param int identificador del depósito con mayor demanada insatisfecha
     * @return double urgencia del cliente
     * Retorna la urgencia del cliente 
     **/
	@Override
	public double getUrgency(int idCustomer, ArrayList<Integer> listIDDepots, NumericMatrix urgencyMatrix, int muIDDepot) {
		double urgency = 0.0;
		int posCustomerMatrix = -1;
		double muDist = 0.0;
		double closestDist = 0.0;
		int posMUDepotMatrix = -1;
		int posDepotMatrixClosest = -1;
		
		posCustomerMatrix = Problem.getProblem().getPosElement(idCustomer);
		posDepotMatrixClosest = Problem.getProblem().getPosElement(listIDDepots.get(0));
		closestDist = urgencyMatrix.getItem(posDepotMatrixClosest, posCustomerMatrix);
		
		posMUDepotMatrix = Problem.getProblem().getPosElement(muIDDepot);
		muDist = urgencyMatrix.getItem(posMUDepotMatrix, posCustomerMatrix);

		if(muDist == Double.POSITIVE_INFINITY) // y la otra distancia no puede ser posiinf
			urgency = closestDist;
		else
			urgency = calculateUrgency(closestDist, muDist);
		
		return urgency;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/classical/urgency/sweep.py
import numpy as np
from typing import List
from .by_urgency import ByUrgency
from .i_urgency_with_mu import IUrgencyWithMU
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class Sweep(ByUrgency, IUrgencyWithMU):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_id_depots: List[int]
        
    def to_clustering(self):
        self.initialize()
        self.assign()
        return self.finish()
        
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_id_depots = list(Problem.get_problem().get_list_id_depots())
        
    def assign(self):
        urgency_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        closest_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        
        self.list_depots_ordered = self.get_depots_ordered(self.list_customers_to_assign, self.list_id_depots, closest_matrix)
        self.mu_id_depot: int = self.find_cluster_with_mu(self.list_clusters)
        self.list_urgencies = self.get_list_urgencies(self.list_customers_to_assign, self.list_depots_ordered, urgency_matrix, self.mu_id_depot)

        while any(self.list_customers_to_assign) and self.list_clusters:
            pos_customer = self.list_urgencies.index(max(self.list_urgencies))
            customer = Problem.get_problem().get_customers()[pos_customer]
            id_customer = customer.get_id_customer()
            request_customer = customer.get_request_customer()

            id_closest_depot = self.list_depots_ordered[pos_customer][0]
            selected_depot: Depot = Problem.get_problem().get_depot_by_id_depot(id_closest_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(selected_depot)
            
            pos_cluster = self.find_cluster(id_closest_depot, self.list_clusters)
            
            if pos_cluster != -1:
                cluster: Cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()
                
                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(id_customer)

                    capacity_depot -= request_customer
                    
                    self.list_customers_to_assign[pos_customer] = None
                    self.list_urgencies[pos_customer] = -1.0
                    self.list_depots_ordered[pos_customer] = None

                    # Recalcular la urgencia si es necesario
                    if id_closest_depot == self.mu_id_depot:
                        self.mu_id_depot = self.find_cluster_with_mu(self.list_clusters)

                        if id_closest_depot != self.mu_id_depot:
                            recalculate_urgency_matrix = self.initialize_cost_matrix(Problem.get_problem().get_customers(), Problem.get_problem().get_depots(), self.distance_type)
                            list_urgencies = self.get_list_urgencies(self.list_customers_to_assign, self.list_depots_ordered, recalculate_urgency_matrix, self.mu_id_depot)
                else:
                    if capacity_depot > request_cluster:
                        self.list_depots_ordered[pos_customer].pop(0)
                        
                        if not self.list_depots_ordered[pos_customer]:
                            self.solution.get_unassigned_items().append(id_customer)
                            
                            self.list_customers_to_assign[pos_customer] = None
                            self.list_urgencies[pos_customer] = -1.0
                            self.list_depots_ordered[pos_customer] = None
                        else:
                            list_urgencies[pos_customer] = self.get_urgency(id_customer, self.list_depots_ordered[pos_customer], urgency_matrix, self.mu_id_depot)
                    else:
                        pos_depot = self.list_id_depots.index(id_closest_depot)

                        for i, depots_ordered in enumerate(self.list_depots_ordered):
                            if id_closest_depot in depots_ordered:
                                depots_ordered.remove(id_closest_depot)
                                if not depots_ordered:
                                    self.solution.get_unassigned_items().append(self.list_customers_to_assign[i].get_id_customer())
                                    
                                    self.list_customers_to_assign[i] = None
                                    self.list_urgencies[i] = -1.0
                                    self.list_depots_ordered[i] = None
                                else:
                                    list_urgencies[i] = self.get_urgency(self.list_customers_to_assign[i].get_id_customer(), depots_ordered, urgency_matrix, self.mu_id_depot)

                        self.list_id_depots.pop(pos_depot)

                        if self.list_clusters[pos_cluster].get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                        else:
                            self.list_clusters.pop(pos_cluster)

    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())
            
        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
                        
        OSRMService.clear_distance_cache()
        
        return self.solution
    
    # Retorna el identificador del depósito cuyo cluster es el de mayor demanada insatisfecha.
    def find_cluster_with_mu(self, clusters: List[Cluster]) -> int:
        id_depot = clusters[0].get_id_cluster()
        mu_request = Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().get_depot_by_id_depot(id_depot)) - clusters[0].get_request_cluster()

        for cluster in clusters[1:]:
            cu_request = Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().get_depot_by_id_depot(cluster.get_id_cluster())) - cluster.get_request_cluster()

            if cu_request > mu_request:
                mu_request = cu_request
                id_depot = cluster.get_id_cluster()
        return id_depot
    
    # Método que retorna un listado con las urgencias de los clientes del listado entrado por parámetro.
    def get_list_urgencies(self, list_customers_to_assign: List[Customer], list_id_depots: List[List[int]], urgency_matrix: np.ndarray, mu_id_depot: int) -> List[float]:
        urgencies: List[float] = []

        if len(list_id_depots) > 1:
            for i, customer in enumerate(list_customers_to_assign):
                if customer is not None:
                    urgencies.append(self.get_urgency(customer.get_id_customer(), list_id_depots[i], urgency_matrix, mu_id_depot))
        else:
            for i, customer in enumerate(list_customers_to_assign):
                if customer is not None:
                    urgencies.append(self.get_urgency(customer.get_id_customer(), list_id_depots[0], urgency_matrix, mu_id_depot))
        return urgencies
    
    # Implementacion del método encargado de obtener la urgencia.
    def get_urgency(self, id_customer: int, list_id_depots: List[int], urgency_matrix: np.ndarray, mu_id_depot: int) -> float:
        urgency: float = 0.0
        closest_dist: float = 0.0
        mu_dist: float = 0.0
        
        total_customers = Problem.get_problem().get_total_customers()
        pos_matrix_customer = Problem.get_problem().get_pos_element(id_customer)
        pos_depot_matrix_closest = Problem.get_problem().get_pos_element(list_id_depots[0])
        
        closest_dist = urgency_matrix[(pos_depot_matrix_closest - total_customers), pos_matrix_customer]

        pos_mu_depot_matrix = Problem.get_problem().get_pos_element(mu_id_depot)
        
        mu_dist = urgency_matrix[(pos_mu_depot_matrix - total_customers), pos_matrix_customer]

        if mu_dist == float('inf'):
            urgency = closest_dist
        else:
            urgency = self.calculate_urgency(closest_dist, mu_dist)

        return urgency